crates/gateway-types/src/transaction_hash.rs

```diff
@@ -0,0 +1,622 @@
+        {
+            let mut hh = HashChain::default();
+            hh = txn.calldata.iter().fold(hh, |mut hh, call_param| {
+                hh.update(call_param.0);
+                hh
+            });
+            hh.finalize()
+        },
+        Some(txn.max_fee),
+        chain_id,
+        txn.nonce,
+        None,
+    )
+}
+
+/// Computes l1 handler transaction hash based on [this formula](https://docs.starknet.io/documentation/architecture_and_concepts/L1-L2_Communication/messaging-mechanism/#structure_and_hashing_l1-l2):
+/// ```text=
+/// l1_handler_tx_hash = h("l1_handler", version, contract_address,
+///     entry_point_selector, h(calldata), 0, chain_id, nonce)
+/// ```
+///
+/// FIXME: SW should fix the formula in the docs
+///
+/// Where `h` is [Pedersen hash](https://docs.starknet.io/documentation/architecture_and_concepts/Hashing/hash-functions/#pedersen_hash)
+///
+/// ## Important
+///
+/// Guarantees correct computation for Starknet **0.9.1** transactions onwards
+fn compute_l1_handler_hash(
+    txn: &L1HandlerTransaction,
+    chain_id: ChainId,
+) -> Result<StarknetTransactionHash> {
+    let call_params_hash = {
+        let mut hh = HashChain::default();
+        hh = txn.calldata.iter().fold(hh, |mut hh, call_param| {
+            hh.update(call_param.0);
+            hh
+        });
+        hh.finalize()
+    };
+
+    let h = compute_txn_hash(
+        b"l1_handler",
+        txn.version,
+        txn.contract_address,
+        Some(txn.entry_point_selector),
+        call_params_hash,
+        None,
+        chain_id,
+        txn.nonce,
+        None,
+    )?;
+
+    if h == txn.transaction_hash {
+        Ok(h)
+    } else {
+        legacy_compute_txn_hash(
+            // Oldest L1 Handler transactions were actually Invokes
+            // which later on were "renamed" to be the former,
+            // yet the hashes remain, hence the prefix
+            b"invoke",
+            txn.contract_address,
+            Some(txn.entry_point_selector),
+            call_params_hash,
+            chain_id,
+        )
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+enum NonceOrClassHash {
+    Nonce(TransactionNonce),
+    ClassHash(ClassHash),
+    None,
+}
+
+impl From<TransactionNonce> for NonceOrClassHash {
+    fn from(n: TransactionNonce) -> Self {
+        Self::Nonce(n)
+    }
+}
+
+impl From<ClassHash> for NonceOrClassHash {
+    fn from(c: ClassHash) -> Self {
+        Self::ClassHash(c)
+    }
+}
+
+impl From<()> for NonceOrClassHash {
+    fn from(_: ()) -> Self {
+        Self::None
+    }
+}
+
+/// _Generic_ compute transaction hash for older transactions (pre 0.8-ish)
+fn legacy_compute_txn_hash(
+    prefix: &[u8],
+    address: ContractAddress,
+    entry_point_selector: Option<EntryPoint>,
+    list_hash: Felt,
+    chain_id: ChainId,
+) -> Result<StarknetTransactionHash> {
+    let mut h = HashChain::default();
+    h.update(Felt::from_be_slice(prefix).context("Converting prefix into felt")?);
```

Comment:
Since `prefix` is an internal constant string can't we simply convert this `?` into an `.expect()` and remove `Result` from the return type?

That would make it possible to return a `VerifyResult` instead of `Result<VerifyResult>` in the top-level `verify()` which I thing would make more sense?